File: wjsquddn/jbw/segmenttree/2042.cpp
#include <cmath>
#include <iostream>
using namespace std;
long long *tree;
long long arr[1000001];
long long init(int node, int start, int end) {
    if (start == end) return tree[node] = arr[start];
    int mid = (start + end) / 2;
    return tree[node] = init(2 * node, start, mid) + init(2 * node + 1, mid + 1, end);
}
long long sum(int node, int start, int end, int left, int right) {
    if (left > end || right < start) return 0;
    if (left <= start && end <= right) return tree[node];
    int mid = (start + end) / 2;
    return sum(node * 2, start, mid, left, right) + sum(node * 2 + 1, mid + 1, end, left, right);
}
void update(int node, int start, int end, int idx, long long diff) {
    if (idx < start || idx > end) return;
    tree[node] += diff;
    if (start == end) return;
    int mid = (start + end) / 2;
    update(node * 2, start, mid, idx, diff);
    update(node * 2 + 1, mid + 1, end, idx, diff);
}
int main() {
    int n, m, k;
    cin >> n >> m >> k;
    int height = ceil(log2(n));
    tree = new long long[1 << (height + 1)];
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    init(1, 0, n - 1);
    int a, b;
    long long c;
    for (int i = 0; i < m + k; i++) {
        cin >> a >> b >> c;
        if (a == 1) {
            long long diff = c - arr[b - 1];
            arr[b - 1] = c;
            update(1, 0, n - 1, b - 1, diff);
        } else {
            cout << sum(1, 0, n - 1, b - 1, c - 1) << endl;
        }
    }
}

File: dalcom/2042/2042.py
from sys import stdin as input

input = open("/Volumes/Program/ps-study/dalcom/2042/input.txt", "rt")

n, m, k = map(int, input.readline().split())

arr = []

for _ in range(n):
    arr.append(int(input.readline()))

tree = [0 for _ in range(len(arr) * 4 + 1)]


def init(start, end, index):
    if start == end:
        tree[index] = arr[start]
        return tree[index]

    mid = (start + end) // 2
    tree[index] = init(start, mid, index * 2) + init(mid + 1, end, index * 2 + 1)
    return tree[index]


def sum(start, end, index, left, right):
    if left > end or right < start:
        return 0

    if left <= start and right >= end:
        return tree[index]

    mid = (start + end) // 2
    return sum(start, mid, index * 2, left, right) + sum(
        mid + 1, end, index * 2 + 1, left, right
    )


def update(start, end, index, arr_index, diff):
    if arr_index < start or arr_index > end:
        return

    tree[index] += diff

    if start == end:
        return

    mid = (start + end) // 2
    update(start, mid, index * 2, arr_index, diff)
    update(mid + 1, end, index * 2 + 1, arr_index, diff)


init(0, len(arr) - 1, 1)

for _ in range(m + k):
    a, b, c = map(int, input.readline().split())
    b = b - 1
    if a == 1:
        diff = c - arr[b]
        arr[b] = c
        update(0, len(arr) - 1, 1, b, diff)
    elif a == 2:
        c = c - 1
        print(sum(0, len(arr) - 1, 1, b, c))
